### PR TITLE
cluster_objects: add preliminary support for cluster objects.

### PR DESCRIPTION
**Pull request by pwhelan**
The new commands added are:

- get cluster_objects: used to get all the cluster objects
- update pipeline_cluster_objects [--pipeline PIPELINE_KEY] [--cluster_object CLUSTER_OBJECT_KEY]: used to enable cluster logging on a namespace.

# Summary of this proposal

Add support for cluster objects to the CLI. 

## Notes for the reviewer

@nicolasparada may implement some APIs to link pipelines to core instances to improve some of the auto-completion and name usage code.

## Issue(s) number(s)

This is related to calyptia/cloud#791 and will be used to implement calyptia/cloud-e2e#42.

## Checklist for submitter

- [ ] My PR has a related issue/bug number.
- [ ] My PR provides tests
  - [ ] Integration tests are added/passes
  - [ ] Unit tests are added/passes
  - [ ] End-to-end tests added/passes
  - [ ] Static code check added/passes
  - [ ] Linting on documentation added/passes
  - [ ] Does not affect code coverage stats
- [ ] My PR requires updating dependencies
- [ ] My PR has the documentation changes required.

## Checklist for reviewer

- [ ] The proposal fixes a bug/issue or implements a new feature that is well described.
- [ ] The proposal has sufficient test cases that covers the changes.
  - [ ] If changes an API, it does not break backwards compatibility (-1 major version).
- [ ] If integration is required, the proposal has integration tests.
- [ ] The proposal does not break coverage stats
- [ ] The proposal has the required documentation changes.

## Backport

- [ ] Backport to the latest stable release.

### CODE REVIEW

File: cmd/calyptia/get_cluster_objects.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+func (config *config) loadClusterObjectID(key string, environmentID string) (string, error) {
+	aa, err := config.fetchAllClusterObjects()
+	if err != nil {
+		return "", err
+	}
+
+	for _, a := range aa {
+		if a.Name == key {
```
Comment (nicolasparada) [inline cmd/calyptia/get_cluster_objects.go:118]: This is not the correct way to do it.
There could be more than one cluster object with the same name on the project; considering the name is only unique per core instance.

Take a look at `loadPipelineID` for example.

File: cmd/calyptia/update_pipeline_cluster_objects.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	fs := cmd.Flags()
+	fs.StringVar(&pipelineKey, "pipeline", "", "Parent pipeline ID or name")
+	fs.StringVar(&clusterObjectKey, "cluster_object", "", "The cluster object ID or Name")
```
Comment (nicolasparada) [inline cmd/calyptia/update_pipeline_cluster_objects.go:51]:
```suggestion
	fs.StringVar(&clusterObjectKey, "cluster-object", "", "The cluster object ID or Name")
```

For the most part, we have been using dash-case for the flags.

File: cmd/calyptia/get_cluster_objects.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+				return yaml.NewEncoder(cmd.OutOrStdout()).Encode(co.Items)
+			default:
+				return fmt.Errorf("unknown output format %q", outputFormat)
+			}
+			return nil
+		},
+	}
+
+	fs := cmd.Flags()
+	fs.StringVar(&coreInstanceKey, "core_instance", "", "Core Instance to list cluster objects from")
```
Comment (nicolasparada) [inline cmd/calyptia/get_cluster_objects.go:190]:
```suggestion
	fs.StringVar(&coreInstanceKey, "core-instance", "", "Core Instance to list cluster objects from")
```

For the most part, we have been using dash-case for the flags.

**Review by nicolasparada — CHANGES_REQUESTED** [review]
There is an issue in the way you load the cluster object ID from a key (name or id).
Also some flags should be in dash-case and not snake_case.

File: cmd/calyptia/get_cluster_objects.go (hunk 1 above)
Comment (pwhelan) [reply cmd/calyptia/get_cluster_objects.go:118]: Is there an API to retrieve the ID of a cluster object by name?

Right now I am working on modifying clusterObjectKeys to return all unique names with their IDs.

File: cmd/calyptia/delete_pipeline_cluster_objects.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
+package main
+
+import (
+	"github.com/spf13/cobra"
+)
+
+func newCmdDeletePipelineClusterObjects(config *config) *cobra.Command {
```
Comment (nicolasparada) [inline cmd/calyptia/delete_pipeline_cluster_objects.go:7]: This method is actually deleting a single cluster object.

```suggestion
func newCmdDeletePipelineClusterObject(config *config) *cobra.Command {
```

File: cmd/calyptia/delete_pipeline_cluster_objects.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	var pipelineKey string
+	var clusterObjectKey string
+	var environment string
+	var encrypt bool
+
+	cmd := &cobra.Command{
+		Use:   "pipeline_cluster_objects",
```
Comment (nicolasparada) [inline cmd/calyptia/delete_pipeline_cluster_objects.go:14]:
```suggestion
		Use:   "pipeline_cluster_object",
```

File: cmd/calyptia/delete_pipeline_cluster_objects.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+		Short: "Delete pipeline cluster objects",
```
Comment (nicolasparada) [inline cmd/calyptia/delete_pipeline_cluster_objects.go:15]:
```suggestion
		Short: "Delete pipeline cluster object",
```

File: cmd/calyptia/update_pipeline_cluster_objects.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
+package main
+
+import (
+	"github.com/spf13/cobra"
+
+	cloud "github.com/calyptia/api/types"
+)
+
+func newCmdUpdatePipelineClusterObjects(config *config) *cobra.Command {
```
Comment (nicolasparada) [inline cmd/calyptia/update_pipeline_cluster_objects.go:9]: Same, this is updating a single cluster object.

```suggestion
func newCmdUpdatePipelineClusterObject(config *config) *cobra.Command {
```

File: cmd/calyptia/update_pipeline_cluster_objects.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	var pipelineKey string
+	var clusterObjectKey string
+	var environment string
+	var encrypt bool
+
+	cmd := &cobra.Command{
+		Use:   "pipeline_cluster_objects",
```
Comment (nicolasparada) [inline cmd/calyptia/update_pipeline_cluster_objects.go:16]:
```suggestion
		Use:   "pipeline_cluster_object",
```

File: cmd/calyptia/update_pipeline_cluster_objects.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+		Short: "Update a file from a pipeline by its name",
```
Comment (nicolasparada) [inline cmd/calyptia/update_pipeline_cluster_objects.go:17]:
```suggestion
		Short: "Update a cluster object from a pipeline by its name or ID.",
```

File: cmd/calyptia/update_pipeline_cluster_objects.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+			if err != nil {
+				return err
+			}
+
+			clusterObjectID, err := config.loadClusterObjectID(clusterObjectKey, environmentID)
+			if err != nil {
+				return err
+			}
+
+			err = config.cloud.UpdatePipelineClusterObjects(config.ctx, pipelineID, cloud.UpdatePipelineClusterObjects{
```
Comment (nicolasparada) [inline cmd/calyptia/update_pipeline_cluster_objects.go:38]: Really, this doesn't update a cluster object, but de/attaches objects to a pipeline... 🤔

File: cmd/calyptia/get_cluster_objects.go [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+	}
+
+	objs := clusterObjectsUniqueByName(aa)
+	for _, obj := range objs {
+		if obj.Name == key {
+			return obj.ID, nil
+		}
+	}
+
+	return "", fmt.Errorf("unable to find unique key")
```
Comment (nicolasparada) [inline cmd/calyptia/get_cluster_objects.go:146]: Still this method is not returning correctly. I think.

Comment (nicolasparada) [reply cmd/calyptia/get_cluster_objects.go:146]: Like, it's getting all cluster objects from all core instances. Then filters all of them that have a unique name, then gets the one that matches the name.
This función then only works by passing a name. But doesn't work if you pass an ID.

File: cmd/calyptia/update_pipeline_cluster_objects.go (hunk 10 above)
Comment (pwhelan) [reply cmd/calyptia/update_pipeline_cluster_objects.go:38]: It indeed does.

File: cmd/calyptia/get_cluster_objects.go (hunk 11 above)
Comment (pwhelan) [reply cmd/calyptia/get_cluster_objects.go:146]: I forgot to add a call to validUUID to return it as an ID. Should I check that before or after checking it as a name?

**Review by nicolasparada — APPROVED** [review]
Thanks.

**Review by patrick-stephens — APPROVED** [review]

**Pull request merged**